node/src/transition_frontier/candidate/transition_frontier_candidate_state.rs

```diff
@@ -89,113 +71,189 @@ impl TransitionFrontierCandidateStatus {
     pub fn is_pending(&self) -> bool {
         matches!(self, Self::SnarkVerifyPending { .. })
     }
-
-    pub fn compared_with(&self) -> Option<&StateHash> {
-        match self {
-            Self::ShortRangeForkResolve { compared_with, .. } => compared_with.as_ref(),
-            _ => None,
-        }
-    }
 }
 
 #[derive(Serialize, Deserialize, Debug, Clone)]
 pub struct TransitionFrontierCandidateState {
-    pub block: Arc<MinaBlockBlockStableV2>,
+    pub block: ArcBlockWithHash,
     pub status: TransitionFrontierCandidateStatus,
     pub chain_proof: Option<(Vec<StateHash>, ArcBlockWithHash)>,
 }
 
+impl Ord for TransitionFrontierCandidateState {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        if self.eq(other) {
+            return std::cmp::Ordering::Equal;
+        }
+        let is_candidate_better = consensus_take(
+            self.block.consensus_state(),
+            other.block.consensus_state(),
+            self.block.hash(),
+            other.block.hash(),
+        );
+        match is_candidate_better {
+            true => std::cmp::Ordering::Less,
+            false => std::cmp::Ordering::Greater,
+        }
+    }
+}
+
+impl PartialOrd for TransitionFrontierCandidateState {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Eq for TransitionFrontierCandidateState {}
+
+impl PartialEq for TransitionFrontierCandidateState {
+    fn eq(&self, other: &Self) -> bool {
+        self.block.hash() == other.block.hash()
+    }
+}
+
 impl TransitionFrontierCandidateState {
     pub fn height(&self) -> u32 {
-        self.block
-            .header
-            .protocol_state
-            .body
-            .consensus_state
-            .blockchain_length
-            .0
-             .0
+        self.block.height()
     }
 }
 
 #[derive(Serialize, Deserialize, Debug, Clone, Default)]
 pub struct TransitionFrontierCandidatesState {
-    pub blocks: BTreeMap<StateHash, TransitionFrontierCandidateState>,
-    // TODO(binier): rename to best candidate. Best tip will be in transition_frontier state.
-    pub best_tip: Option<StateHash>,
-    pub best_tip_chain_proof: Option<(Vec<StateHash>, ArcBlockWithHash)>,
+    /// Maintains an ordered list of transition frontier Candidates,
+    /// ordered using consensus rules worst to best.
+    ordered: BTreeSet<TransitionFrontierCandidateState>,
+    /// Invalid candidates.
+    invalid: BTreeMap<StateHash, u32>,
```

Comment:
would it be worth it to add the reason of the invalidation?

Reply:
those that fail validations. I'll expand on it

Reply:
Oh I didn't mean the comment, but to keep information about what was the reason for the failure (block too early/late? invalid proof? invalid protocol version? etc)

Reply:
Btw, from the prevalidation errors there are two that are temporary and recoverable: `GenesisNotReady` and `ReceivedTooEarly`